Newsletter categories: Render non-interactive newsletter category pills in the editor

TimBroddin opened the pull request:
Closes https://github.com/Automattic/wp-calypso/issues/80904

## Proposed changes:
- This adds the newsletter categories as pills in the editor.

<img width="682" alt="CleanShot 2023-08-24 at 16 23 51@2x" src="https://github.com/Automattic/jetpack/assets/528287/c91c2dc3-3bc3-400f-a35c-220c89c36bed">

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
For completeness, please test this on WPCOM & Jetpack self hosted.

1. Checkout this PR and run the command below for WPCOM or build it and run it though JT
2. Mark some categories as newsletter categories: `POST /wpcom/v2/sites/<siteid>/newsletter-categories/<categoryid>` 
3. Create a new post/page & add the subscribe block
4. The newsletter categories should be above the input field

TimBroddin committed:
[not verified] Add newsletter categories to subscription block

Added a new feature to the subscription block which displays newsletter categories. The added categories are styled and displayed above the subscription form. Also added a new hook to query newsletter categories from the backend. This feature enhances user experience by offering categorized subscribing options. The form width is reduced when categories are displayed for better aesthetics.

TimBroddin committed:
[not verified] Add newsletter categories to subscription block

Added a new feature to the subscription block which displays newsletter categories. The added categories are styled and displayed above the subscription form. Also added a new hook to query newsletter categories from the backend. This feature enhances user experience by offering categorized subscribing options. The form width is reduced when categories are displayed for better aesthetics.

phcp commented on projects/plugins/jetpack/extensions/blocks/subscriptions/view.scss:
```diff
@@ -1,3 +1,5 @@
+
+
```
Could you remove those extra lines?

phcp commented on projects/plugins/jetpack/extensions/blocks/subscriptions/hooks/use-newsletter-categories.js:
```diff
@@ -0,0 +1,33 @@
+		fetchData();
+	}, [] );
+
+	return { data, enabled: data.length, error, loading };
```
Just confirm, `enabled: data.length` means that the API should reflect the user choice on the Newsletter Settings page (to enable/disable newsletter categories), right?

phcp replied:
I'm thinking of a scenario where the user enabled newsletter categories, selected a few of them, and saved; then the user decided to disable newsletter categories. In the current implementation, the selected categories are still there, so the API would return categories.

TimBroddin replied:
You're right. In that case we need to extend the API endpoint to expose this setting.

TimBroddin replied:
I added the extra field in D119783-code & updated the code of this PR to use it.

TimBroddin committed:
[not verified] Add 'enabled' state to newsletter category hooks

The changes introduced a new 'enabled' state to the useNewsletterCategories hook. This was done to allow better manipulation of the subscription blocks based on whether the newsletter category function is enabled by the user or not, hence improving user experience when interacting with subscription blocks.

TimBroddin committed:
[not verified] Remove unnecessary spacing in view.scss

This commit removes the additional two lines of spacing at the beginning of the view.scss file within the subscriptions block. This was done to clean the code and make it adhering to the standard coding practices, which enhances the readability and maintainability.

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/hooks/use-newsletter-categories.js:
```diff
@@ -0,0 +1,35 @@
+import apiFetch from '@wordpress/api-fetch';
+import { select } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+import { CONNECTION_STORE_ID } from '../../../../../../js-packages/connection';
```
Here it may be best to import from the published package instead:

```suggestion
import { CONNECTION_STORE_ID } from '@automattic/jetpack-connection';
import apiFetch from '@wordpress/api-fetch';
import { select } from '@wordpress/data';
import { useEffect, useState } from '@wordpress/element';
```

Better yet, since all we need here is the blog ID, we should be able to get it from the data passed in PHP, with `window?.Jetpack_Editor_Initial_State?.wpcomBlogId`

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/hooks/use-newsletter-categories.js:
```diff
@@ -0,0 +1,35 @@
+		const fetchData = async () => {
+			try {
+				const newsLetterCategories = await apiFetch( {
+					url: `https://public-api.wordpress.com/wpcom/v2/sites/${ blogId }/newsletter-categories`,
```
We've typically not queried `public-api` directly in our blocks. Instead we query a local endpoint that in turn queries WordPress.com (or makes a direct request when we're on WordPress.com).  The local endpoints live in `projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/`. 

There is some documentation about it here:

https://github.com/Automattic/jetpack/blob/695768a21b1341684a4da54fa4abc39093cd7ed8/projects/plugins/jetpack/extensions/README.md#L262-L271

TimBroddin replied:
Thanks for the pointers once again @jeherve .

I've made the needed API changes: https://github.com/Automattic/jetpack/pull/32666/commits/596bcef21f4fc3f89fc44768f728e72f86cede54

I do have a question regarding defusion: When I test with the jetpack-downloader, I can see the endpoint file on my WP.COM Simple site. However, even after deleting the existing endpoint, I cannot get it to execute it using the WP.com console. Do files inside `wpcom-endpoints` not get loaded on Simple Sites?

It's not a problem for now, since everything works. Just interested.

jeherve replied:
> Do files inside wpcom-endpoints not get loaded on Simple Sites?

Yeah, it's a bit confusing. You'd need to do something like D110282-code (check the last file being added, or other files in that directory for some more examples).

TimBroddin replied:
Thank you! It's much cleared now. I'll create a separate patch for WPCOM, removing the endpoint, after the Jetpack release happens on WPCOM.

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/view.scss:
```diff
@@ -96,3 +96,32 @@
+		font-size: 14px;
+		font-weight: 500;
+		line-height: 20px;
+		color: var(--studio-gray-50);;
+		border: 1px solid var(--studio-gray-5);
+		border-radius: var(--border-radius);
```
I wonder if we could use WordPress colors here, for more consistency?

Here are some examples:
https://github.com/Automattic/jetpack/blob/3a3c4adc1c0913bac224ddb832713bd6101fe1ab/projects/plugins/jetpack/extensions/blocks/donations/view.scss#L28

TimBroddin committed:
[not verified] Add newsletter categories REST API endpoint and adjust client-side fetch

This commit introduces a new REST API endpoint for retrieving the newsletter categories. The new endpoint will enable a more seamless and effective integration of these categories in the client-side.

Moreover, modifications were also made to the 'trait-wpcom-rest-api-proxy-request-trait.php' file. Now the body for Client::wpcom_json_api_request_as_user is checked if not null. This was done to prevent an error when the body is null.

Lastly, the fetch request at 'use-newsletter-categories.js' has been updated to use the new /newsletter-categories endpoint within the local server environment instead of making a direct request to the public API.

TimBroddin committed:
[not verified] Update styling & endpoint

jeherve reviewed (changes requested):
Noting that you'll need to look at the failing js tests in `extensions/blocks/subscriptions/test/edit.js` before we can merge this:
https://github.com/Automattic/jetpack/actions/runs/6010883027/job/16303218143?pr=32666

TimBroddin committed:
[not verified] Add newsletter categories mock function for testing

In the subscriptions module, a mock function was added in the test/edit.js file. This was done to simulate the behavior of the actual 'useNewsletterCategories' function in the unit tests. This will facilitate the testing process by controlling its response and help in improving the code coverage.

TimBroddin commented:
> Noting that you'll need to look at the failing js tests in `extensions/blocks/subscriptions/test/edit.js` before we can merge this: https://github.com/Automattic/jetpack/actions/runs/6010883027/job/16303218143?pr=32666

Fixed it! And somehow managed to bring in all the labels while doing a rebase 🫠

jeherve commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-newsletter-categories-list.php:
```diff
@@ -0,0 +1,91 @@
+	 */
+	public function get_newsletter_categories() {
+
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM && function_exists( 'require_lib' ) ) {
```
Since you're doing a `( new Host() )->is_wpcom_simple()` check above, maybe we have an opportunity to create a new method in this class to check for wpcom?

TimBroddin replied:
I guess I can just leave out this check, since this callback will only be invoked on WPCOM.

jeherve commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-newsletter-categories-list.php:
```diff
@@ -0,0 +1,91 @@
+				$this,
+				'get_newsletter_categories',
+			) : array( $this, 'proxy_request_to_wpcom_as_user' ),
+			'permission_callback' => '__return_true',
```
Should we restrict access to folks who can manage options on the site?

jeherve commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-newsletter-categories-list.php:
```diff
@@ -0,0 +1,91 @@
+ * REST API endpoint for the Jetpack Blogroll block.
+ *
+ * @package automattic/jetpack
+ * @since 12.2
```
```suggestion
 * @since $$next-version$$
```

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/view.scss:
```diff
@@ -96,3 +98,32 @@
+
+	.wp-block-jetpack-subscriptions__newsletter-category {
+		padding: 6px 10px;
+		font-family: "SF Pro Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen-Sans, Ubuntu, Cantarell, "Helvetica Neue", sans-serif;
```
We may be able to use more variables here, like 

```suggestion
		font-family: $default-font;
```

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/view.scss:
```diff
@@ -96,3 +98,32 @@
+		font-weight: 500;
+		line-height: 20px;
+		color: $gray-600;
+		border: 1px solid #dcdcde;
```
Could this be a `$gray-200`?

```suggestion
		border: 1px solid $gray-200;
```

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/view.scss:
```diff
@@ -96,3 +98,32 @@
+	.wp-block-jetpack-subscriptions__newsletter-category {
+		padding: 6px 10px;
+		font-family: "SF Pro Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen-Sans, Ubuntu, Cantarell, "Helvetica Neue", sans-serif;
+		font-size: 14px;
```
Could we use `$default-font-size` here?

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/edit.js:
```diff
@@ -179,6 +183,10 @@ export function SubscriptionEdit( props ) {
 	};
 
+	const cssVars = {
+		'--border-radius': borderRadius ? borderRadius + 'px' : DEFAULT_BORDER_RADIUS_VALUE + 'px',
```
I wonder if we would need to be more specific with the variable name here, to avoid potential mixups later?

TimBroddin committed:
- Better usage of CSS variables
- Renamed --border-radius to --subscribe-block-border-radius
- Removed check of WPCOM

phcp reviewed (approved):
Good job @TimBroddin! Tested and it works as expected.

TimBroddin commented:
Hey @jeherve 

The changes you asked for were made, could you rereview? 🙏

jeherve reviewed (approved):
Looks good and tests well for me. 👍

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/subscription-block-add-newsletter-categories
